File: apps/app/src/api/curseforge.rs
use crate::api::Result;
use tauri::plugin::TauriPlugin;
use tauri::Runtime;

#[tauri::command]
pub async fn search_curseforge(
    query: &str,
    class_id: Option<u32>,
    game_version: Option<String>,
    mod_loader_type: Option<u32>,
    page: Option<u32>,
    page_size: Option<u32>,
    sort_field: Option<u32>,
    sort_order: Option<String>,
) -> Result<theseus::curseforge::SearchResponse> {
    Ok(theseus::curseforge::search_curseforge(
        query.to_string(),
        class_id,
        game_version,
        mod_loader_type,
        page,
        page_size,
        sort_field,
        sort_order,
    ).await?)
}

#[tauri::command]
pub async fn get_mod_curseforge(mod_id: u32) -> Result<theseus::curseforge::CurseForgeMod> {
    Ok(theseus::curseforge::get_mod_cf(mod_id).await?)
}

#[tauri::command]
pub async fn get_mod_files_curseforge(
    mod_id: u32,
    game_version: Option<String>,
    mod_loader_type: Option<u32>,
) -> Result<Vec<theseus::curseforge::CurseForgeFile>> {
    Ok(theseus::curseforge::get_mod_files_cf(mod_id, game_version, mod_loader_type).await?)
}

pub fn init<R: Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::<R>::new("curseforge")
        .invoke_handler(tauri::generate_handler![
            search_curseforge,
            get_mod_curseforge,
            get_mod_files_curseforge,
        ])
        .build()
}

File: apps/app-frontend/src/helpers/curseforge.ts
import { invoke } from '@tauri-apps/api/core'

export interface CurseForgeMod {
	id: number
	name: string
	summary: string
	logo?: { thumbnail_url: string; url: string }
	authors: { name: string; url: string }[]
	download_count: number
	date_modified: string
	categories: { id: number; name: string; url: string }[]
}

export interface CurseForgeFile {
	id: number
	mod_id: number
	display_name: string
	file_name: string
	download_url: string | null
	file_date: string
	game_versions: string[]
	dependencies: { mod_id: number; relation_type: number }[]
}

export async function search_cf(
	query: string,
	class_id?: number,
	game_version?: string,
	mod_loader_type?: number,
	page?: number,
) {
	return await invoke<any>('plugin:curseforge|search_curseforge', {
		query,
		classId: class_id,
		gameVersion: game_version,
		modLoaderType: mod_loader_type,
		page,
	})
}

export async function get_mod_cf(modId: number) {
	return await invoke<CurseForgeMod>('plugin:curseforge|get_mod_curseforge', { modId })
}

export async function get_mod_files_cf(
	modId: number,
	gameVersion?: string,
	modLoaderType?: number,
) {
	return await invoke<CurseForgeFile[]>('plugin:curseforge|get_mod_files_curseforge', {
		modId,
		gameVersion,
		modLoaderType,
	})
}
